--- swupd-client-3.18.5/src/search.c~	2018-11-27 23:28:17.000000000 +0000
+++ swupd-client-3.18.5/src/search.c	2018-12-14 17:52:32.591576628 +0000
@@ -41,7 +41,7 @@
 
 static char search_type = '0';
 static char scope = '0';
-static int num_results = INT_MAX;
+static int num_results = 5;
 
 /* bundle_result contains the information to print along with
  * the internal relevancy score used to sort the output */
