diff --git a/src/curl.c b/src/curl.c
index 924e6f4..f347ed0 100644
--- a/src/curl.c
+++ b/src/curl.c
@@ -446,15 +446,6 @@ static CURLcode swupd_curl_set_security_opts(CURL *curl)
 		goto exit;
 	}
 
-	curl_ret = curl_easy_setopt(curl, CURLOPT_PINNEDPUBLICKEY, UPDATE_CA_CERTS_PATH "/425b0f6b.key");
-	if (curl_ret != CURLE_OK) {
-		goto exit;
-	}
-
-	curl_ret = curl_easy_setopt(curl, CURLOPT_CAPATH, UPDATE_CA_CERTS_PATH);
-	if (curl_ret != CURLE_OK) {
-		goto exit;
-	}
 
 #if 0
 	// TODO: add the below when you know the paths:
