diff --git a/src/scripts.c b/src/scripts.c
index 6c0a12e..7e36f42 100644
--- a/src/scripts.c
+++ b/src/scripts.c
@@ -82,14 +82,10 @@ void run_scripts(void)
 	/* path_prefix aware helper */
 	if (need_update_boot) {
 		update_kernel();
-	} else {
-		printf("No kernel update needed, skipping helper call out.\n");
 	}
 
 	if (need_update_bootloader) {
 		update_bootloader();
-	} else {
-		printf("No bootloader update needed, skipping helper call out.\n");
 	}
 
 	/* helpers which don't run when path_prefix is set */
